### ISSUE TITLE
TypeError: Cannot read property 'length' of undefined - when using .error()

### ISSUE THREAD

**Issue by wlechowicz** [AndrewKeig/express-validation#59]
When trying to set a custom error message as per [Joi API reference](https://github.com/hapijs/joi/blob/v10.2.2/API.md#anyerrorerr), for example:
`Joi.string().error(new Error('expected a string, actually'))`
then express-validation throws
`TypeError: Cannot read property 'length' of undefined` in index.js line 76
because `.error(new Error('message'))` produces an Error instance without `.details` or other decorations of the `ValidationError` so `errors.details.length` cannot be accessed directly as `errors.details` is undefined.

Should another route be chosen to allow for custom error messages (especially important for regexes) or should the condition be improved to distinguish between ValidationErrors and generic Errors?

**Comment by wlechowicz** [comment]
Since Joi asserts if the argument provided to the `.error()` function is an instance of Error, then one quite ugly workaround is to do 
```
class CustomError extends Error {
  constructor(message, field) {
    super(message);
    this.details = [{ message: message, path: field }];
  }
}
```
and then in schema
```
body: {
  ...
  subpath: {
    ...
    some_field: Joi.string().error(new CustomError('I really wanted a string', 'subpath.some_field'))
  }
}
```
which will produce an output such as 
```
"errors": [
    {
      "field": "subpath.some_field",
      "location": "body",
      "messages": [
        "I really wanted a string"
      ],
      "types": [
        null
      ]
    }
  ]
```
which kind of works, but again, is ugly and bad and redundant (you have to type the path manually, because at the time of evaluating `.error()`, it's not populated by Joi yet).

**Comment by colthreepv** [comment]
Okay, that was unexpected.. never tried this functionality.
I think you want to have a custom error message from Joi, instead of the default one

If you wish to keep the Joi error structure, you can try using the  `.options()` functionality

Example:
```
Joi.string().options({
  language: {
    string: { base: 'This is not a string, please give me a string!!!' }
  }
 }))
```

If you read carefully the documentation, overriding these messages is quite tricky, as they tend to be as specific as possible.

express-validation objective is giving an easy way to produce "standard" validation errors, so supporting custom errors it's kind of out of scope
What I would suggest you is to use plain Joi in your project if you aim to do so... or is there any other use case?

**Comment by wlechowicz** [comment]
Actually a simple string isn't a very good example of the issue I reported, I just used it for simplicity (and mentioned that actual use case is for regexes). Most common case where you'd want to override an error message on a schema entry level basis would be regex validated strings, so for
```
field_name: Joi.string().regex(/^[a-zA-Z0-9\\\s\-\.]*$/)
```
you get something more user-friendly (or frontend-friendly) than
```
"\"field_name\" with value \"&#261;\" fails to match the required pattern: /^[a-zA-Z0-9\\\\\\s\\-\\.]*$/"
``` 
as this is not something you'd want displayed under a form input :)
This is done with .error() and cannot be done with language options which are global overrides.

**Comment by colthreepv** [comment]
There is a simple option to name a regex "pattern", I know that is not optimal.

Still, what's the alternative?
I only see few options:

  * `express-validation` checks if a "custom" error is returned from Joi, and just forwards it.
     Now you will have what you expect, but what really happens it's this library will output 99% of the time a standard ValidationError, and when Joi is overridden, returns an arbitrary, different, structure.

  * `express-validation` wraps the custom error making it compatible in the actual object structure using a CustomError implementation.
    It requires ugly and redundant code in the user side, that not only needs to know about Joi and express-validation, but also needs to write ad-hoc code to make them both work

  * Excalate the issue to Joi, that should let users give a custom error message, without changing the error structure
    This seems the most long sighted approach to me, there is a [old but gold issue related](https://github.com/hapijs/joi/issues/546) that might come to help

If you have any other ideas please suggest, I will think about it as well

**Comment by wlechowicz** [comment]
I agree that the root of the evil is that Joi chooses to have a completely different outcome whenever `.error()` is used, I'm just arguing that it would make sense to safeguard against that and not have `express-validation` blow up whenever someone uses `.error(new Error('I love giving my users a 500 response'))` in the schema.

**Comment by colthreepv** [comment]
Sure, I see your point, probably will go with that (1), the custom branch should be a little more solid now.

I will see how that merges in the 2.x branch, will release it also for 1.x version to avoid application crashes

**Comment by AndrewKeig** [comment]
Closing, released 3.0
